test/net-test.js

```diff
@@ -0,0 +1,2218 @@
+        const peer = Peer.fromOptions({spv: true});
+        peer.outbound = true;
+
+        const pkt = new packets.VersionPacket({
+          services: 0 | services.NETWORK
+        });
+
+        let err = null;
+
+        try {
+          await peer.handleVersion(pkt);
+        } catch (e) {
+          err = e;
+        }
+        assert(err);
+        assert.equal(err.message,
+                     'Peer does not support BIP37.');
+      });
+
+      it('will error if w/o bloom version (outbound)', async () => {
+        const peer = Peer.fromOptions({spv: true});
+        peer.outbound = true;
+
+        const pkt = new packets.VersionPacket({
+          services: 0 | services.NETWORK | services.BLOOM,
+          version: common.BLOOM_VERSION - 1
+        });
+
+        let err = null;
+
+        try {
+          await peer.handleVersion(pkt);
+        } catch (e) {
+          err = e;
+        }
+        assert(err);
+        assert.equal(err.message,
+                     'Peer does not support BIP37.');
+      });
+
+      it('will error if w/o witness service (outbound)', async () => {
+        const peer = Peer.fromOptions({});
+        peer.outbound = true;
+
+        const pkt = new packets.VersionPacket({
+          services: 0 | services.NETWORK
+        });
+
+        let err = null;
+
+        try {
+          await peer.handleVersion(pkt);
+        } catch (e) {
+          err = e;
+        }
+        assert(err);
+        assert.equal(err.message,
+                     'Peer does not support segregated witness.');
+      });
+
+      it('will send ack (outbound)', async () => {
+        const peer = Peer.fromOptions({});
+        peer.outbound = true;
+
+        const pkt = new packets.VersionPacket({
+          services: 0 | services.NETWORK | services.WITNESS
+        });
+
+        let called = false;
+
+        peer.send = (packet) => {
+          assert(packet);
+          assert.equal(packet.type, packets.types.VERACK);
+          called = true;
+        };
+
+        await peer.handleVersion(pkt);
+        assert(called);
+      });
+
+      it('will send ack (outbound=false)', async () => {
+        const peer = Peer.fromOptions({});
+        peer.outbound = false;
+
+        const pkt = new packets.VersionPacket();
+
+        let called = false;
+
+        peer.send = (packet) => {
+          assert(packet);
+          assert.equal(packet.type, packets.types.VERACK);
+          called = true;
+        };
+
+        await peer.handleVersion(pkt);
+        assert(called);
+      });
+    });
+
+    describe('handleVerack', function() {
+      it('will set ack', async () => {
+        const peer = Peer.fromOptions({});
+        assert.equal(peer.ack, false);
+        const pkt = new packets.VerackPacket();
+        await peer.handleVerack(pkt);
+        assert.equal(peer.ack, true);
+      });
+    });
+
+    describe('handlePing', function() {
+      it('will not send ping without nonce', async () => {
```

Comment:
Do you mean `will not send PONG...` ?

Reply:
Yes, was a typo as "piong" and should have been fixed as "pong".